### ISSUE TITLE
Possible false positive: `Rc::ptr_eq` and `vtable_address_comparisons`

### ISSUE THREAD

**Issue by hunger** [rust-lang/rust-clippy#7461]
I probably just do not understand the problem, so sorry for this report!

I have two variables `a` and `b` holding a `Rc<dyn Foo>`.  I do `Rc::ptr_eq(&a, &b)` and the `vtable_address_comparisons` lint is triggered, which seems wrong to me.

Lint name: vtable_address_comparisons

I tried this code:

```rust
use std::rc::Rc;

struct Baz(String);

trait Foo {
    fn bar(&self);
}

impl Foo for Baz {
    fn bar(&self) { println!("{}", self.0); }
}

fn main() {
    let a: Rc<dyn Foo> = Rc::new(Baz("2".to_string()));
    let b: Rc<dyn Foo> = Rc::new(Baz("1".to_string()));
    
    assert!(!Rc::ptr_eq(&a, &b));
}
```

and got this clippy lint:

```
error: comparing trait object pointers compares a non-unique vtable address
  --> src/main.rs:17:14
   |
17 |     assert!(!Rc::ptr_eq(&a, &b));
   |              ^^^^^^^^^^^^^^^^^^
   |
   = note: `#[deny(clippy::vtable_address_comparisons)]` on by default
   = help: consider extracting and comparing data pointers only
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#vtable_address_comparisons
```

Is that correct? I do not understand where the vtable comes into the picture here!

`Rc<T>` seems to only hold a pointer to a `RcBox<T>`. That `RcBox<T>` in turn holds two reference counts and the actual `T` the `Rc<T>` manages. So I would expect (in the case above) for a fat pointer to be in the `RcBox`, while the `Rc` itself holds a "normal" pointer.

`Rc::ptr_eq` compares that normal pointer in `Rc` itself as far as I can tell, which in my limited understanding should be fine. So why does clippy warn about this?

Where is my mistake? I am sure I made one;-) Clippy usually has such wonderful reports and explanations, so I would appreciate some more guidance in this lint.

# Meta:

- Playground gives the clippy version as `0.1.55 (2021-07-12 955b9c0)`
- `rustc` from the rust playground just now.

Thank you for your work on clippy! It is such a wonderful tool!

**Comment by ogoffart** [comment]
I think the lint is valid.
If i understand correctly, this is how we would reproduce. (i did not test this)

```
// crate1
fn foo(xx: Rc<u32>) -> Rc<dyn Display> { xx }
// crate2
fn foo(xx: Rc<u32>) -> Rc<dyn Display> { xx }

// crate3
let xx = Rc::new(42);
assert!(Rc::ptr_eq(crate1::foo(xx.clone()), crate2::foo(xx.clone())));
```

If i understand the problem correctly here is that the vtable pointer to Display might not be unique in both crate for some reason and so the assert could fail.

**Comment by hunger** [comment]
@ogoffart: Yes, I get that the vtable for `Display` may be different.

But `Rc` holds a pointer `ptr` in `Rc` that points to an `RcBox<T>`. That is a normal pointer for all I can tell.

That `RcBox` in turn holds a `T`, which is `dyn Display` in our case.

`Rc::ptr_eq` compares the `ptr` in `Rc` as far as I understand the code. That is a thin pointer and should be OK to compare. I do not get why clippy warns about a fat pointer comparison -- yes, such a fat pointer is hidden in the object pointer to by the pointer `Rc::ptr` is looking at, but that should be safe, shouldn't it?

But as I said before: I am probably missing something crucial!

**Comment by tronical** [comment]
I think that you might be missing is that `Rc<T>` is the size of the pointer to the `RcBox<T>` but `Rc<dyn Foo>` - in order to work - is the size of two pointers: The pointer to the allocation (`RcBox<T>`) as well as a pointer to the vtable. That way it's possible that an `Rc<T>`, `Rc<dyn Foo>` as well as an `Rc<dyn Bar>` can share the same reference count and allocation of `T` implements also `Foo` and `Bar`.

See also https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ce41fa1cdf9df28502ba4fb96019ba49 for an example

**Comment by hunger** [comment]
@tronical: Aaaahhh! Now that makes sense!

This obviously also explains nicely why the clippy lint is generated in this place.

So as suspected I misunderstood the code.
